display/man.rs

```diff
@@ -0,0 +1,445 @@
+        .spawn()?;
+
+    if let Some(stdin) = stdin {
+        if let Some(mut process_stdin) = process.stdin.take() {
+            process_stdin.write_all(stdin)?;
+        } else {
+            return Err(io::Error::new(
+                io::ErrorKind::Other,
+                format!("failed to open stdin for {name}"),
+            ));
+        }
+    }
+
+    let output = process.wait_with_output().map_err(|_| {
+        io::Error::new(io::ErrorKind::Other, format!("failed to get {name} stdout"))
+    })?;
+
+    if !output.status.success() {
+        Err(io::Error::new(
+            io::ErrorKind::Other,
+            format!("{name} failed"),
+        ))
+    } else {
+        Ok(output)
+    }
+}
+
+/// Gets system documentation content by passed name.
+///
+/// # Arguments
+///
+/// `name` - [str] name of necessary system documentation.
+///
+/// # Returns
+///
+/// [Vec<u8>] output of called `*cat` command.
+///
+/// # Errors
+///
+/// [std::io::Error] if file not found or failed to execute `*cat` command.
+fn get_man_page(name: &str) -> Result<Vec<u8>, io::Error> {
+    let man_page_path = get_man_page_path(name)?;
+
+    let cat_process_name = if man_page_path.extension().and_then(|ext| ext.to_str()) == Some("gz") {
+        "zcat"
+    } else {
+        "cat"
+    };
+
+    spawn(cat_process_name, &[man_page_path], None, Stdio::piped()).map(|output| output.stdout)
+}
+
+/// Gets page width.
+///
+/// # Returns
+///
+/// [Option<u16>] width value of current terminal. [Option::Some] if working on terminal and receiving terminal size was succesfull. [Option::None] if working not on terminal.
+///
+/// # Errors
+///
+/// Returns [ManError] if working on terminal and failed to get terminal size.
+fn get_page_width() -> Result<Option<u16>, ManError> {
+    if std::io::stdout().is_terminal() {
+        let mut winsize = libc::winsize {
+            ws_row: 0,
+            ws_col: 0,
+            ws_xpixel: 0,
+            ws_ypixel: 0,
+        };
+        let result = unsafe { libc::ioctl(libc::STDOUT_FILENO, libc::TIOCGWINSZ, &mut winsize) };
+        if result == 0 {
+            let result_width = if winsize.ws_col >= 80 {
+                winsize.ws_col - 2
+            } else {
+                winsize.ws_col
+            };
+            Ok(Some(result_width))
+        } else {
+            Err(ManError("failed to get terminal width".to_string()))
+        }
+    } else {
+        Ok(None)
+    }
+}
+
+/// Gets formated by `groff(1)` system documentation.
+///
+/// # Arguments
+///
+/// `man_page` - [&[u8]] with content that needs to be formatted.
+/// `width` - [Option<u16>] width value of current terminal.
+///
+/// # Returns
+///
+/// [Vec<u8>] STDOUT of called `groff(1)` formatter.
+///
+/// # Errors
+///
+/// [std::io::Error] if file failed to execute `groff(1)` formatter.
+fn groff_format(man_page: &[u8], width: Option<u16>) -> Result<Vec<u8>, io::Error> {
+    let mut args = vec![
+        "-Tutf8".to_string(),
```

Comment:
Why do args have to be owned?

Reply:
After this array is defined, I have `if let ...` statement where I need to push owned string after `format!(...)`.

Reply:
```rust
fn groff_format(man_page: &[u8], width: Option<u16>) -> Result<Vec<u8>, io::Error> {
    let mut args = vec![
        "-Tutf8",
        "-S",
        "-P-h",
        "-Wall",
        "-mtty-char",
        "-t",
        "-mandoc",
    ];
    let width = width.map(|w| (format!("-rLL={w}n"), format!("-rLR={w}n")));
    if let Some((rll, rlr)) = width.as_ref() {
        args.push(rll);
        args.push(rlr);
    }

    spawn("groff", &args, Some(man_page), Stdio::piped()).map(|output| output.stdout)
}
```

Reply:
Thanks!